python/ray/serve/storage/checkpoint_path.py

```diff
@@ -0,0 +1,62 @@
+from urllib.parse import parse_qsl, urlparse
+
+from ray.serve.utils import logger, import_from_string
+from ray.serve.constants import DEFAULT_CHECKPOINT_PATH
+from ray.serve.storage.kv_store import (RayInternalKVStore, RayLocalKVStore,
+                                        RayS3KVStore)
+from ray.serve.storage.kv_store_base import KVStoreBase
+
+
+def make_kv_store(checkpoint_path, namespace):
+    """Create KVStore instance based on checkpoint_path configuration"""
+
+    print(namespace)
+
+    if checkpoint_path == DEFAULT_CHECKPOINT_PATH:
+        logger.info("Using RayInternalKVStore for controller "
+                    "checkpoint and recovery.")
+        return RayInternalKVStore(namespace)
+    else:
+        parsed_url = urlparse(checkpoint_path)
+        if parsed_url.scheme not in {"s3", "file", "custom"}:
+            raise ValueError(
+                f"Checkpoint must be one of `{DEFAULT_CHECKPOINT_PATH}`, "
+                "`file://path...`, `s3://path...` or "
+                "`custom://my_modue:ClassName?arg1=val1`. But it is "
+                f"{checkpoint_path}")
+
+        if parsed_url.scheme == "file":
+            db_path = parsed_url.netloc + parsed_url.path
+            logger.info("Using RayLocalKVStore for controller "
+                        f"checkpoint and recovery: path={db_path}")
+            return RayLocalKVStore(namespace, db_path)
+
+        if parsed_url.scheme == "s3":
+            bucket = parsed_url.netloc
+            prefix = parsed_url.path
+            logger.info(
+                "Using Ray S3 KVStore for controller checkpoint and recovery: "
+                f"bucket: {bucket}, checkpoint path: {checkpoint_path}")
+            return RayS3KVStore(
+                namespace,
+                bucket=bucket,
+                prefix=prefix,
+            )
+
+        if parsed_url.scheme == "custom":
+            kwargs = dict(parse_qsl(parsed_url.query))
+
+            # Prepare the parameters to initialize imported class.
+            checkpoint_provider = parsed_url.netloc
+            KVStoreClass = import_from_string(checkpoint_provider)
```

Comment:
why do we need to support this ?

Reply:
so it's pluggable and users can write their own internal storage plugins, like HTTP based, SQL based, or checkpoint to Azure etc.